### PR TITLE
Refactor scheduler

### PR DESCRIPTION
**Pull request by Lockna**
In this pull request the whole scheduler architecture is rewritten.
Instead of depending on interrupts to schedule new thread I went for a different approach, where the context gets switched by just swapping some callee-saved registers and switching stacks.
In addition to the new scheduling architecture a timer-event infrastructure is introduced.
For various reasons DPC's also will be implemented in this pull request do have a good interaction with the timer-events.

### CODE REVIEW

**Comment by Lockna** [comment]
Still TODO:
- Remove dead code in scheduler.rs
- Maybe keep the scheduler struct, which holds run_queue, current_thread and everything else needed for MLFQ. Or do not make a scheduler struct and just some loose functions which work on the queues in current_cpu.

**Comment by Lockna** [comment]
Fix working on release build

**Comment by Lockna** [comment]
FIXME: Reschedule dpc takes away execution. Should not be the case

File: xernel/kernel/src/arch/amd64/interrupts/mod.rs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    let next_free_vector = free_vectors[ipl as usize];
+
+    if next_free_vector > base_vector + 15 {
+        return None;
+    }
+
+    free_vectors[ipl as usize] += 1;
+
+    Some(next_free_vector)
+}
```
Comment (0x6D70) [inline xernel/kernel/src/arch/amd64/interrupts/mod.rs:121]: This function does not check whether the assigned vector was already set in the ``init`` function

File: xernel/kernel/src/arch/amd64/interrupts/mod.rs [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+pub fn register_handler(vector: u8, handler: fn(&mut TrapFrame)) {
+    let mut handlers = INTERRUPT_HANDLERS.lock();
+
+    match handlers[vector as usize] {
+        IRQHandler::None => {}
+        _ => unreachable!("register_handler: handler has already been registered"),
```
Comment (0x6D70) [inline xernel/kernel/src/arch/amd64/interrupts/mod.rs:128]: this is not unreachable, use something else ...

File: xernel/kernel/src/arch/amd64/ioapic.rs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    pub unsafe fn write(&mut self, reg: u32, val: u32) {
+        ((self.address) as *mut u32).write_volatile(reg);
+        ((self.address + 0x10) as *mut u32).write_volatile(val);
+    }
+
+    pub unsafe fn mask_irq(&mut self) {}
+
+    pub unsafe fn unmask_irq(&mut self) {}
```
Comment (0x6D70) [inline xernel/kernel/src/arch/amd64/ioapic.rs:29]: Add ``todo!``

File: xernel/kernel/src/cpu.rs [hunk 4]
```diff
@@ -100,44 +106,70 @@ impl<T> PerCpu<T> {
...
+    fn deref(&self) -> &Self::Target {
+        self.get()
+    }
 }
 
-impl PerCpuData {
-    pub fn set_kernel_stack(&mut self, kernel_stack: usize) {
-        self.kernel_stack = kernel_stack;
-    }
+#[repr(C)]
```
Comment (0x6D70) [inline xernel/kernel/src/cpu.rs:117]: this should be ``packed``

File: xernel/kernel/src/sched/scheduler.rs [hunk 5]
```diff
@@ -1,255 +1,107 @@
...
-    let mut thread = thread.lock();
+pub fn switch_threads(old: Arc<Thread>, new: Arc<Thread>) {
+    old.status.set(ThreadStatus::Ready);
 
-    thread.status = ThreadStatus::Running;
+    new.status.set(ThreadStatus::Running);
 
-    if !thread.is_kernel_thread() {
+    if !new.is_kernel_thread() {
+        println!("i am here where i should not be");
```
Comment (0x6D70) [inline xernel/kernel/src/sched/scheduler.rs:70]: remove ...

File: xernel/kernel/src/sched/scheduler.rs [hunk 6]
```diff
@@ -1,255 +1,107 @@
...
-    APIC.eoi();
-    APIC.create_oneshot_timer(*SCHEDULER_VECTOR, thread.priority.ms() * 1000);
+    *current_cpu().current_thread.write() = Some(new.clone());
 
-    thread.unlock();
-    sched.unlock();
-
-    restore_context(context);
+    unsafe {
+        // FIXME: If println is used after some time page fault happens
```
Comment (0x6D70) [inline xernel/kernel/src/sched/scheduler.rs:94]: This seems to work now

**Review by 0x6D70 — CHANGES_REQUESTED** [review]
Good work :+1:

File: xernel/kernel/src/arch/amd64/interrupts/mod.rs (hunk 1 above)
Comment (Lockna) [reply xernel/kernel/src/arch/amd64/interrupts/mod.rs:121]: Currently not used. Will be rewritten when actually used

**Pull request merged**